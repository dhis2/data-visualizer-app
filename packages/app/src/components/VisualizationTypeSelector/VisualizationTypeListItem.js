import React from 'react'
import PropTypes from 'prop-types'

import ListItemIcon from './ListItemIcon'

import styles from './styles/VisualizationTypeSelector.module.css'

const VisualizationTypeListItem = ({
    iconType,
    label,
    disabled,
    isSelected,
    onClick,
}) => {
    const classNames = [styles.listItem]

    if (isSelected) {
        classNames.push(styles.listItemActive)
    }

    if (disabled) {
        classNames.push(styles.listItemDisabled)
    }

    return (
        <div className={classNames.join(' ')} onClick={onClick}>
            <span className={styles.listItemIcon}>
                {
                    <ListItemIcon
                        iconType={iconType}
                        style={{ width: 48, height: 48 }}
                    />
                }
            </span>
            <span className={styles.listItemText}>{label}</span>
        </div>
    )
}

VisualizationTypeListItem.propTypes = {
    disabled: PropTypes.bool,
    iconType: PropTypes.string,
    isSelected: PropTypes.bool,
    label: PropTypes.string,
    onClick: PropTypes.func,
}

export default VisualizationTypeListItem
