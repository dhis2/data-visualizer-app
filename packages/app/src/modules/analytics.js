export const computeGenericPeriodNames = responses => {
    const xAxisRes = responses.reduce((out, res) => {
        if (out.metaData) {
            if (
                res.metaData.dimensions.pe.length >
                out.metaData.dimensions.pe.length
            ) {
                out = res;
            }
        } else {
            out = res;
        }

        return out;
    }, {});

    const metadata = xAxisRes.metaData;

    return metadata.dimensions.pe.reduce((genericPeriodNames, periodId) => {
        const name = metadata.items[periodId].name;

        // until the day the backend will support this in the API:
        // trim off the trailing year in the period name
        // english names should all have the year at the end of the string
        genericPeriodNames.push(name.replace(/\s+\d{4}$/, ''));

        return genericPeriodNames;
    }, []);
};
