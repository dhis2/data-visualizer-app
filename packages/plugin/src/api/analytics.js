const peId = 'pe'

export const apiFetchAnalytics = async (d2, current, options) => {
    const req = new d2.analytics.request()
        .fromModel(current)
        .withParameters(options)

    const rawResponse = await d2.analytics.aggregate.get(req)

    return [new d2.analytics.response(rawResponse)]
}

export const apiFetchAnalyticsForYearOverYear = async (
    d2,
    current,
    options
) => {
    let yearlySeriesReq = new d2.analytics.request()
        .addPeriodDimension(current.yearlySeries)
        .withSkipData(true)
        .withSkipMeta(false)
        .withIncludeMetadataDetails(true)

    if (options.relativePeriodDate) {
        yearlySeriesReq = yearlySeriesReq.withRelativePeriodDate(
            options.relativePeriodDate
        )
    }

    const yearlySeriesRes = await d2.analytics.aggregate.fetch(yearlySeriesReq)

    const requests = []
    const yearlySeriesLabels = []

    const now = new Date()
    const currentDay = ('' + now.getDate()).padStart(2, 0)
    const currentMonth = ('' + (now.getMonth() + 1)).padStart(2, 0)

    yearlySeriesRes.metaData.dimensions[peId].forEach(period => {
        yearlySeriesLabels.push(yearlySeriesRes.metaData.items[period].name)

        const startDate = `${period}-${currentMonth}-${currentDay}`

        const req = new d2.analytics.request()
            .fromModel(current)
            .withParameters(options)
            .withRelativePeriodDate(startDate)

        requests.push(d2.analytics.aggregate.get(req))
    })

    return Promise.all(requests).then(responses => ({
        responses: responses.map(res => new d2.analytics.response(res)),
        yearlySeriesLabels,
    }))
}
