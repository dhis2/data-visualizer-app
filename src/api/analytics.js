import {
    Analytics,
    layoutGetDimensionItems,
    VIS_TYPE_PIVOT_TABLE,
    DIMENSION_ID_DATA,
    DIMENSION_ID_PERIOD,
    DAILY,
    WEEKLY,
    WEEKS_THIS_YEAR,
} from '@dhis2/analytics'
import {
    METHOD_MODIFIED_Z_SCORE,
    METHOD_STANDARD_Z_SCORE,
    OUTLIER_METHOD_PROP,
    OUTLIER_THRESHOLD_PROP,
} from '../components/VisualizationOptions/Options/OutliersForOutlierTable.js'
import { OUTLIER_MAX_RESULTS_PROP } from '../components/VisualizationOptions/Options/OutliersMaxResults.js'
import {
    getRelativePeriodTypeUsed,
    getOutlierTableDimensionIdHeaderMap,
} from '../modules/analytics.js'
import { getSortingFromVisualization } from '../modules/ui.js'

const periodId = DIMENSION_ID_PERIOD

export const apiFetchAnalytics = async (dataEngine, visualization, options) => {
    const analyticsEngine = Analytics.getAnalytics(dataEngine)

    const req = new analyticsEngine.request()
        .fromVisualization(visualization)
        .withParameters(options)
        .withIncludeNumDen(visualization.type === VIS_TYPE_PIVOT_TABLE)

    const rawResponse = await analyticsEngine.aggregate.get(req)

    return [new analyticsEngine.response(rawResponse)]
}

export const getAnalyticsRequestForOutlierTable = ({
    analyticsEngine,
    visualization,
    options,
    forDownload = false,
}) => {
    const dimensionIdHeaderMap = getOutlierTableDimensionIdHeaderMap(options)

    const parameters = {
        ...options,
        maxResults: visualization.outlierAnalysis[OUTLIER_MAX_RESULTS_PROP],
        algorithm:
            visualization.outlierAnalysis[OUTLIER_METHOD_PROP] ===
            METHOD_STANDARD_Z_SCORE
                ? 'Z_SCORE'
                : visualization.outlierAnalysis[OUTLIER_METHOD_PROP],
        threshold: visualization.outlierAnalysis[OUTLIER_THRESHOLD_PROP],
    }

    const columns = visualization.columns || []
    const headers = []

    columns.forEach(({ dimension, items }) => {
        parameters[dimension] = items.map(({ id }) => id).join(',')

        headers.push(forDownload ? dimension : dimensionIdHeaderMap[dimension])

        if (dimension === DIMENSION_ID_DATA) {
            headers.push('cocname')
        }
    })

    headers.push('value')

    switch (visualization.outlierAnalysis.outlierMethod) {
        case METHOD_MODIFIED_Z_SCORE:
            headers.push('median', 'modifiedzscore', 'medianabsdeviation')
            break
        case METHOD_STANDARD_Z_SCORE:
            headers.push('mean', 'zscore', 'stddev')
    }

    headers.push('lowerbound', 'upperbound')

    parameters.headers = headers.join(',')

    // sorting
    const sorting = getSortingFromVisualization(visualization)

    if (sorting) {
        parameters.orderBy = sorting.dimension
        parameters.sortOrder = sorting.direction
    }

    return new analyticsEngine.request().withParameters(parameters)
}

export const apiFetchAnalyticsForOutlierTable = async (
    dataEngine,
    visualization,
    options
) => {
    const analyticsEngine = Analytics.getAnalytics(dataEngine)

    const req = getAnalyticsRequestForOutlierTable({
        analyticsEngine,
        visualization,
        options,
    })

    const rawResponse = await analyticsEngine.aggregate.getOutliersData(req)

    return [new analyticsEngine.response(rawResponse)]
}

export const apiFetchAnalyticsForYearOverYear = async (
    dataEngine,
    visualization,
    options
) => {
    const analyticsEngine = Analytics.getAnalytics(dataEngine)

    let yearlySeriesReq = new analyticsEngine.request()
        .addPeriodDimension(visualization.yearlySeries)
        .withSkipData(true)
        .withSkipMeta(false)
        .withIncludeMetadataDetails(true)

    if (options.relativePeriodDate) {
        yearlySeriesReq = yearlySeriesReq.withRelativePeriodDate(
            options.relativePeriodDate
        )
    }

    const yearlySeriesRes = await analyticsEngine.aggregate.fetch(
        yearlySeriesReq
    )

    const periodDates = []
    const yearlySeriesLabels = []

    const now = new Date()
    const currentDay = ('' + now.getDate()).padStart(2, 0)
    const currentMonth = ('' + (now.getMonth() + 1)).padStart(2, 0)

    const periodItems = layoutGetDimensionItems(visualization, periodId)

    // relative week period in use
    if (
        getRelativePeriodTypeUsed(periodItems) === WEEKLY &&
        !periodItems[0].id === WEEKS_THIS_YEAR
    ) {
        const relativeWeeksData = await prepareRequestsForRelativeWeeks({
            analyticsEngine,
            visualization,
            options,
            yearlySeriesRes,
            currentMonth,
            currentDay,
        })

        periodDates.push(...relativeWeeksData.periodDates)
        yearlySeriesLabels.push(...relativeWeeksData.yearlySeriesLabels)
    } else if (getRelativePeriodTypeUsed(periodItems) === DAILY) {
        const relativeDaysData = prepareRequestsForRelativeDays({
            yearlySeriesRes,
            currentMonth,
            currentDay,
        })

        periodDates.push(...relativeDaysData.periodDates)
        yearlySeriesLabels.push(...relativeDaysData.yearlySeriesLabels)
    } else {
        yearlySeriesRes.metaData.dimensions[periodId]
            .sort()
            .reverse()
            .forEach((year) => {
                yearlySeriesLabels.push(
                    yearlySeriesRes.metaData.items[year].name
                )

                periodDates.push(`${year}-${currentMonth}-${currentDay}`)
            })
    }

    // request analytics data/metaData for each year in the serie with its own specific relativePeriodDate
    const requests = periodDates.reduce((list, periodDate) => {
        const req = new analyticsEngine.request()
            .fromVisualization(visualization)
            .withParameters(options)
            .withRelativePeriodDate(periodDate)

        list.push(analyticsEngine.aggregate.get(req))

        return list
    }, [])

    return Promise.all(requests).then((responses) => ({
        responses: responses.map((res) => new analyticsEngine.response(res)),
        yearlySeriesLabels,
    }))
}

const prepareRequestsForRelativeDays = ({
    yearlySeriesRes,
    currentMonth,
    currentDay,
}) => {
    const yearlySeriesLabels = []
    const periodDates = []

    const yearlySeriesIds = yearlySeriesRes.metaData.dimensions[periodId]
        .slice()
        .sort()
        .reverse()

    const isFeb29 = currentMonth === '02' && currentDay === '29'

    yearlySeriesIds.forEach((year) => {
        yearlySeriesLabels.push(yearlySeriesRes.metaData.items[year].name)

        const isLeapYear = new Date(year, 1, 29).getDate() === 29

        // 1. check if current date is feb 29
        // 2. check if current year is NOT a leap year
        if (isFeb29 && !isLeapYear) {
            // 3. use feb 28 for that year as relativePeriodDate
            periodDates.push(`${year}-02-28`)
        } else {
            periodDates.push(`${year}-${currentMonth}-${currentDay}`)
        }
    })

    return { yearlySeriesLabels, periodDates }
}

// special handling for when a relative weeks period is selected as category
// this takes care of data alignment issues between different years when one of the years have 53 weeks
// and more in general when the returned weeks for different years are not exactly the same range
// and data points must be "shifted" in the right position
// See https://jira.dhis2.org/browse/DHIS2-9729
const prepareRequestsForRelativeWeeks = async ({
    analyticsEngine,
    visualization,
    options,
    yearlySeriesRes,
    currentMonth,
    currentDay,
}) => {
    const yearlySeriesLabels = []
    const periodDates = []

    const yearlySeriesIds = yearlySeriesRes.metaData.dimensions[periodId]
        .slice()
        .sort()
        .reverse()

    // 1. request metadata of last year of the serie (with relativePeriodDate === today)
    // 2. extract the last week number of the LAST_x_WEEKS period
    // 3. request metadata for the same week number for each one of the other years of the serie
    // 3. compute relativePeriodDate for each other year of the serie:
    //    this is done by adding 1 day to the endDate of the week period obtained above
    // 4. request analytics data/metaData for each year in the serie with its own specific relativePeriodDate
    const referencePeriodYear = yearlySeriesIds.shift()

    yearlySeriesLabels.push(
        yearlySeriesRes.metaData.items[referencePeriodYear].name
    )

    periodDates.push(`${referencePeriodYear}-${currentMonth}-${currentDay}`)

    // 1. request metadata of last year of the serie (with relativePeriodDate === today)
    const referencePeriodReq = new analyticsEngine.request()
        .fromVisualization(visualization)
        .withParameters(options)
        .withRelativePeriodDate(
            `${referencePeriodYear}-${currentMonth}-${currentDay}`
        )
        .withSkipData(true)
        .withSkipMeta(false)

    const referencePeriodRes = await analyticsEngine.aggregate.fetch(
        referencePeriodReq
    )

    // 2. extract the last week number of the LAST_x_WEEKS period
    //    special handling for the week 53 case as not all years have 53 weeks
    const referenceWeekPeriod =
        referencePeriodRes.metaData.dimensions[periodId].pop()
    const [referenceWeekYear, referenceWeekNumber] =
        referenceWeekPeriod.split('W')
    const referenceWeekYearDelta = referencePeriodYear - referenceWeekYear

    const weekPeriods = yearlySeriesIds.reduce((periods, year) => {
        yearlySeriesLabels.push(yearlySeriesRes.metaData.items[year].name)

        periods.push(`${year - referenceWeekYearDelta}W${referenceWeekNumber}`)

        // edge case for week 53, not all years have it, so request also week 52
        if (referenceWeekNumber === '53') {
            periods.push(`${year - referenceWeekYearDelta}W52`)
        }

        return periods
    }, [])

    if (weekPeriods.length) {
        // 3. request metadata for the same week number for each one of the other years of the serie
        const weekPeriodsReq = new analyticsEngine.request()
            .addPeriodDimension(weekPeriods)
            .withSkipData(true)
            .withSkipMeta(false)
            .withIncludeMetadataDetails(true)

        const weekPeriodsRes = await analyticsEngine.aggregate.fetch(
            weekPeriodsReq
        )

        // 3. compute relativePeriodDate for each other year of the serie:
        //    this is done by adding 1 day to the endDate of the week period obtained above
        const seenYears = []

        weekPeriodsRes.metaData.dimensions[periodId]
            .sort()
            .reverse()
            .forEach((period) => {
                const year = period.substr(0, 4)

                // make sure we only take W53 or W52 whichever is available
                if (!seenYears.includes(year)) {
                    const periodDate = new Date(
                        weekPeriodsRes.metaData.items[period].endDate
                    )
                    periodDate.setDate(periodDate.getDate() + 1)

                    periodDates.push(
                        `${periodDate.getFullYear()}-${(
                            '' +
                            (periodDate.getMonth() + 1)
                        ).padStart(2, 0)}-${(
                            '' + periodDate.getDate()
                        ).padStart(2, 0)}`
                    )

                    seenYears.unshift(year)
                }
            })
    }

    return { yearlySeriesLabels, periodDates }
}
