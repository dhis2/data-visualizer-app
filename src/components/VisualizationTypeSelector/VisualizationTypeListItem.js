import PropTypes from 'prop-types'
import React from 'react'
import ListItemIcon from './ListItemIcon.js'
import styles from './styles/VisualizationTypeSelector.module.css'

const VisualizationTypeListItem = ({
    iconType,
    label,
    description,
    disabled,
    isSelected,
    onClick,
}) => {
    const classNames = [styles.listItem]

    if (isSelected) {
        classNames.push(styles.listItemActive)
    }

    if (disabled) {
        classNames.push(styles.listItemDisabled)
    }

    return (
        <div className={classNames.join(' ')} onClick={onClick}>
            <div className={styles.listItemIcon}>
                {
                    <ListItemIcon
                        iconType={iconType}
                        style={{ width: 48, height: 48 }}
                    />
                }
            </div>
            <div className={styles.listItemText}>
                <p className={styles.listItemName}>{label}</p>
                <p className={styles.listItemDescription}>{description}</p>
            </div>
        </div>
    )
}

VisualizationTypeListItem.propTypes = {
    description: PropTypes.string,
    disabled: PropTypes.bool,
    iconType: PropTypes.string,
    isSelected: PropTypes.bool,
    label: PropTypes.string,
    onClick: PropTypes.func,
}

export default VisualizationTypeListItem
