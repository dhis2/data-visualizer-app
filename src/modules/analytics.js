import {
    getFixedDimensions,
    getRelativePeriodsOptionsById,
    DIMENSION_ID_DATA,
    DIMENSION_ID_ORGUNIT,
    DIMENSION_ID_PERIOD,
    WEEKLY,
    DAILY,
} from '@dhis2/analytics'
import i18n from '@dhis2/d2-i18n'

export const outlierTableHeadersMap = {
    [DIMENSION_ID_DATA]: 'dxname',
    [DIMENSION_ID_ORGUNIT]: 'ouname',
    [DIMENSION_ID_PERIOD]: 'pename',
}

export const getOutlierTableDimensionIdHeaderMap = ({ showHierarchy }) => ({
    [DIMENSION_ID_DATA]: 'dxname',
    [DIMENSION_ID_ORGUNIT]: showHierarchy ? 'ounamehierarchy' : 'ouname',
    [DIMENSION_ID_PERIOD]: 'pename',
})

export const getOutlierTableHeadersDetails = () => {
    const fixedDimensions = getFixedDimensions()

    return {
        dxname: {
            label: fixedDimensions[DIMENSION_ID_DATA]?.name,
        },
        ouname: {
            label: fixedDimensions[DIMENSION_ID_ORGUNIT]?.name,
        },
        ounamehierarchy: {
            label: fixedDimensions[DIMENSION_ID_ORGUNIT]?.name,
        },
        pename: {
            label: fixedDimensions[DIMENSION_ID_PERIOD]?.name,
        },
        cocname: {
            label: i18n.t('Category option combination'),
        },
        value: {
            label: i18n.t('Value'),
        },
        absdev: {
            label: i18n.t('Absolute deviation'),
            tooltip: i18n.t(
                'A measure of the absolute difference between each data point and a central value, usually the mean or median, providing a straightforward understanding of dispersion in the dataset.'
            ),
        },
        modifiedzscore: {
            label: i18n.t('Modified Z-score'),
            tooltip: i18n.t(
                'A measure of how far a data point deviates from the median, using the median absolute deviation instead of the standard deviation, making it robust against outliers.'
            ),
        },
        median: {
            label: i18n.t('Median'),
            tooltip: i18n.t(
                "The middle value in a dataset when the values are arranged in ascending or descending order. It's a robust measure of central tendency that is less affected by outliers compared to the mean."
            ),
        },
        medianabsdeviation: {
            label: i18n.t('Median absolute deviation'),
            tooltip: i18n.t(
                "A robust measure of variability, found by calculating the median of the absolute differences between each data point and the overall median. It's less influenced by outliers compared to other measures like the standard deviation."
            ),
        },
        zscore: {
            label: i18n.t('Z-score'),
            tooltip: i18n.t(
                'A measure of how many standard deviations a data point is from the mean of a dataset, providing insight into how unusual or typical that data point is relative to the rest of the distribution.'
            ),
        },
        mean: {
            label: i18n.t('Mean'),
            tooltip: i18n.t('Average of the value over time.'),
        },
        stddev: {
            label: i18n.t('Standard deviation'),
            tooltip: i18n.t(
                'A measure of how dispersed the data is in relation to the mean. Low standard deviation indicates data are clustered tightly around the mean, and high standard deviation indicates data are more spread out.'
            ),
        },
        lowerbound: {
            label: i18n.t('Min'),
            tooltip: i18n.t('Minimum score threshold'),
        },
        upperbound: {
            label: i18n.t('Max'),
            tooltip: i18n.t('Maximum score threshold'),
        },
    }
}

export const computeYoYMatrix = (responses, relativePeriodTypeUsed) => {
    const periodGroups = responses.reduce((list, res) => {
        list.push(res.metaData.dimensions.pe)

        return list
    }, [])

    if (relativePeriodTypeUsed === DAILY) {
        periodGroups.sort((a, b) => a[0].substr(-2) - b[0].substr(-2))

        const periodKeyAxisIndexMatrix = periodGroups
            .shift()
            .map((periodId) => [periodId])

        periodGroups.forEach((periodGroup) => {
            periodGroup.forEach((periodId) => {
                const matchGroups = periodId.match(/(\d{4})(\d{2})(\d{2})/)

                const month = matchGroups[2]
                const day = matchGroups[3]

                // find same month/day in 1st "serie"
                const xAxisIndexForPeriod = periodKeyAxisIndexMatrix.findIndex(
                    (periodKeys) => periodKeys[0].substr(4) === `${month}${day}`
                )

                if (xAxisIndexForPeriod !== -1) {
                    periodKeyAxisIndexMatrix[xAxisIndexForPeriod].push(periodId)
                } else if (month === '02' && day === '29') {
                    // February 29 special case
                    // find index for february 28
                    const indexForFeb28 = periodKeyAxisIndexMatrix.findIndex(
                        (periodKeys) =>
                            periodKeys.findIndex((periodKey) =>
                                /0228$/.test(periodKey)
                            ) !== -1
                    )

                    if (indexForFeb28 !== -1) {
                        periodKeyAxisIndexMatrix.splice(indexForFeb28 + 1, 0, [
                            periodId,
                        ])
                    } else {
                        periodKeyAxisIndexMatrix.push([periodId])
                    }
                } else {
                    periodKeyAxisIndexMatrix.push([periodId])
                }
            })
        })

        return periodKeyAxisIndexMatrix
    } else if (relativePeriodTypeUsed === WEEKLY) {
        periodGroups.sort((a, b) => b[0].split('W')[1] - a[0].split('W')[1])

        const periodKeyAxisIndexMatrix = periodGroups
            .shift()
            .map((periodId) => [periodId])

        periodGroups.forEach((periodGroup) => {
            periodGroup.forEach((periodId) => {
                const [year, week] = periodId.split('W')

                // find week number in 1st "serie"
                const xAxisIndexForPeriod = periodKeyAxisIndexMatrix.findIndex(
                    (periodKeys) => periodKeys[0].split('W')[1] === week
                )

                if (xAxisIndexForPeriod !== -1) {
                    periodKeyAxisIndexMatrix[xAxisIndexForPeriod].push(periodId)
                } else if (week === '1') {
                    const indexForW2 = periodKeyAxisIndexMatrix.findIndex(
                        (periodKeys) =>
                            periodKeys.findIndex((periodKey) =>
                                /W2$/.test(periodKey)
                            ) !== -1
                    )

                    if (indexForW2 !== -1) {
                        periodKeyAxisIndexMatrix[indexForW2].push(periodId)
                    } else {
                        periodKeyAxisIndexMatrix.push([periodId])
                    }
                } else {
                    // find the right spot considering also the year
                    const indexForPrevWeekInYear =
                        periodKeyAxisIndexMatrix.findIndex(
                            (periodKeys) =>
                                periodKeys.findIndex(
                                    (periodKey) =>
                                        periodKey === `${year}W${week - 1}`
                                ) !== -1
                        )

                    periodKeyAxisIndexMatrix.splice(
                        indexForPrevWeekInYear + 1,
                        0,
                        [periodId]
                    )
                }
            })
        })

        return periodKeyAxisIndexMatrix
    } else {
        const periodKeyAxisIndexMatrix = periodGroups.reduce(
            (list, periodGroup) => {
                periodGroup.forEach((periodId, index) => {
                    list[index].push(periodId)
                })

                return list
            },
            Array.from({ length: periodGroups[0].length }, () => [])
        )

        return periodKeyAxisIndexMatrix
    }
}

export const computeGenericPeriodNamesFromMatrix = (
    periodKeyAxisIndexMatrix,
    relativePeriodTypeUsed
) => {
    switch (relativePeriodTypeUsed) {
        case WEEKLY:
            return (
                periodKeyAxisIndexMatrix
                    // remove year, return "Wnn"
                    .map((periodKeys) => periodKeys[0].substr(4))
                    .flat()
            )
        case DAILY:
            return periodKeyAxisIndexMatrix
                .map((periodKeys) =>
                    // remove year, return "dd-mm"
                    periodKeys[0].substr(4).replace(/(\d{2})(\d{2})/, '$2-$1')
                )
                .flat()
    }
}

export const computeGenericPeriodNames = (responses) => {
    const xAxisRes = responses.reduce((out, res) => {
        if (out.metaData) {
            if (
                res.metaData.dimensions.pe.length >
                out.metaData.dimensions.pe.length
            ) {
                out = res
            }
        } else {
            out = res
        }

        return out
    }, {})

    const metadata = xAxisRes.metaData

    return metadata.dimensions.pe.reduce((genericPeriodNames, periodId) => {
        const name = metadata.items[periodId].name

        // until the day the backend will support this in the API:
        // trim off the trailing year in the period name
        // english names should all have the year at the end of the string
        genericPeriodNames.push(name.replace(/\s+\d{4}$/, ''))

        return genericPeriodNames
    }, [])
}

export const getRelativePeriodTypeUsed = (periodItems) => {
    if (
        getRelativePeriodsOptionsById(WEEKLY)
            .getPeriods()
            .find((period) => period.id === periodItems[0].id)
    ) {
        return WEEKLY
    } else if (
        getRelativePeriodsOptionsById(DAILY)
            .getPeriods()
            .find((period) => period.id === periodItems[0].id)
    ) {
        return DAILY
    }
}
