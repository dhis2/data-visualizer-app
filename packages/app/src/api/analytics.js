import { getInstance } from 'd2';
import { FIXED_DIMENSIONS } from '../modules/fixedDimensions';

const peId = FIXED_DIMENSIONS.pe.id;

export const apiDownloadImage = async (type, formData) => {
    const d2 = await getInstance();
    const api = d2.Api.getApi();

    const url = `${api.baseUrl}/svg.${type}`;

    return api
        .fetch(url, {
            method: 'POST',
            body: formData,
            headers: new Headers(api.defaultHeaders),
        })
        .then(res => res.blob());
};

export const apiDownloadData = async (current, format, idScheme, path) => {
    const d2 = await getInstance();
    const api = d2.Api.getApi();

    let req = new d2.analytics.request()
        .fromModel(current, path === 'dataValueSet')
        .withFormat(format);

    if (path) {
        req = req.withPath(path);
    }

    if (idScheme) {
        req = req.withOutputIdScheme(idScheme);
    }

    const url = new URL(
        `${api.baseUrl}/${req.buildUrl()}`,
        `${window.location.origin}${window.location.pathname}`
    );

    Object.entries(req.buildQuery()).forEach(([key, value]) =>
        url.searchParams.append(key, value)
    );

    return url;
};

export const apiFetchAnalytics = async (current, options) => {
    const d2 = await getInstance();

    const req = new d2.analytics.request()
        .fromModel(current)
        .withParameters(options);

    const rawResponse = await d2.analytics.aggregate.get(req);

    return [new d2.analytics.response(rawResponse)];
};

export const apiFetchAnalyticsForYearOverYear = async (current, options) => {
    const d2 = await getInstance();

    let yearlySeriesReq = new d2.analytics.request()
        .addPeriodDimension(current.yearlySeries)
        .withSkipData(true)
        .withSkipMeta(false)
        .withIncludeMetadataDetails(true);

    if (options.relativePeriodDate) {
        yearlySeriesReq = yearlySeriesReq.withRelativePeriodDate(
            options.relativePeriodDate
        );
    }

    const yearlySeriesRes = await d2.analytics.aggregate.fetch(yearlySeriesReq);

    const requests = [];
    const yearlySeriesLabels = [];

    const now = new Date();
    const currentDay = ('' + now.getDate()).padStart(2, 0);
    const currentMonth = ('' + (now.getMonth() + 1)).padStart(2, 0);

    yearlySeriesRes.metaData.dimensions[peId].forEach(period => {
        yearlySeriesLabels.push(yearlySeriesRes.metaData.items[period].name);

        const startDate = `${period}-${currentMonth}-${currentDay}`;

        const req = new d2.analytics.request()
            .fromModel(current)
            .withParameters(options)
            .withRelativePeriodDate(startDate);

        requests.push(d2.analytics.aggregate.get(req));
    });

    return Promise.all(requests).then(responses => ({
        responses: responses.map(res => new d2.analytics.response(res)),
        yearlySeriesLabels,
    }));
};
