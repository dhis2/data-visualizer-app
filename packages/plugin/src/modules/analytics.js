import { getRelativePeriodsOptionsById, WEEKS, DAYS } from '@dhis2/analytics'

export const computeYoYMatrix = (responses, relativePeriodTypeUsed) => {
    const periodGroups = responses.reduce((list, res) => {
        list.push(res.metaData.dimensions.pe)

        return list
    }, [])

    if (relativePeriodTypeUsed === WEEKS) {
        periodGroups.sort((a, b) => b[0].split('W')[1] - a[0].split('W')[1])

        const periodKeyAxisIndexMatrix = periodGroups
            .shift()
            .map(periodId => [periodId])

        periodGroups.forEach(periodGroup => {
            periodGroup.forEach(periodId => {
                const [year, week] = periodId.split('W')

                // find week number in 1st "serie"
                const xAxisIndexForPeriod = periodKeyAxisIndexMatrix.findIndex(
                    periodKeys => periodKeys[0].split('W')[1] === week
                )

                if (xAxisIndexForPeriod !== -1) {
                    periodKeyAxisIndexMatrix[xAxisIndexForPeriod].push(periodId)
                } else if (week === '1') {
                    const indexForW2 = periodKeyAxisIndexMatrix.findIndex(
                        periodKeys =>
                            periodKeys.findIndex(periodKey =>
                                /W2$/.test(periodKey)
                            ) !== -1
                    )

                    if (indexForW2 !== -1) {
                        periodKeyAxisIndexMatrix[indexForW2].push(periodId)
                    } else {
                        periodKeyAxisIndexMatrix.push([periodId])
                    }
                } else {
                    // find the right spot considering also the year
                    const indexForPrevWeekInYear = periodKeyAxisIndexMatrix.findIndex(
                        periodKeys =>
                            periodKeys.findIndex(
                                periodKey => periodKey === `${year}W${week - 1}`
                            ) !== -1
                    )

                    periodKeyAxisIndexMatrix.splice(
                        indexForPrevWeekInYear + 1,
                        0,
                        [periodId]
                    )
                }
            })
        })

        return periodKeyAxisIndexMatrix
        // TODO daily case (Feb 29th)
    } else {
        const periodKeyAxisIndexMatrix = periodGroups.reduce(
            (list, periodGroup) => {
                periodGroup.forEach((periodId, index) => {
                    list[index].push(periodId)
                })

                return list
            },
            Array.from({ length: periodGroups[0].length }, () => [])
        )

        return periodKeyAxisIndexMatrix
    }
}

export const computeGenericPeriodNamesFromMatrix = (
    periodKeyAxisIndexMatrix,
    relativePeriodTypeUsed
) => {
    switch (relativePeriodTypeUsed) {
        case WEEKS:
            return (
                periodKeyAxisIndexMatrix
                    // remove year, return "Wnn"
                    .map(periodKeys => periodKeys[0].substr(4))
                    .flat()
            )
        case DAYS:
            return periodKeyAxisIndexMatrix
                .map(periodKeys =>
                    // remove year, return "dd-mm"
                    periodKeys[0].substr(4).replace(/(\d{2})(\d{2})/, '$2-$1')
                )
                .flat()
    }
}

export const computeGenericPeriodNames = responses => {
    const xAxisRes = responses.reduce((out, res) => {
        if (out.metaData) {
            if (
                res.metaData.dimensions.pe.length >
                out.metaData.dimensions.pe.length
            ) {
                out = res
            }
        } else {
            out = res
        }

        return out
    }, {})

    const metadata = xAxisRes.metaData

    return metadata.dimensions.pe.reduce((genericPeriodNames, periodId) => {
        const name = metadata.items[periodId].name

        // until the day the backend will support this in the API:
        // trim off the trailing year in the period name
        // english names should all have the year at the end of the string
        genericPeriodNames.push(name.replace(/\s+\d{4}$/, ''))

        return genericPeriodNames
    }, [])
}

export const getRelativePeriodTypeUsed = periodItems => {
    if (
        getRelativePeriodsOptionsById(WEEKS)
            .getPeriods()
            .find(period => period.id === periodItems[0].id)
    ) {
        return WEEKS
    } else if (
        getRelativePeriodsOptionsById(DAYS)
            .getPeriods()
            .find(period => period.id === periodItems[0].id)
    ) {
        return DAYS
    }
}
