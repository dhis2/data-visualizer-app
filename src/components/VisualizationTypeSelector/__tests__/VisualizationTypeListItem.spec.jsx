import { VIS_TYPE_COLUMN } from '@dhis2/analytics'
import { render } from '@testing-library/react'
import VisualizationTypeListItem from '../VisualizationTypeListItem.jsx'

test('VisualizationTypeListItem renders Column item', () => {
    const props = {
        iconType: VIS_TYPE_COLUMN,
        label: 'Column Chart',
        description: 'A chart that displays data in columns',
        disabled: false,
        isSelected: false,
        onClick: jest.fn(),
    }

    const { container } = render(<VisualizationTypeListItem {...props} />)
    expect(container).toMatchSnapshot()
})
