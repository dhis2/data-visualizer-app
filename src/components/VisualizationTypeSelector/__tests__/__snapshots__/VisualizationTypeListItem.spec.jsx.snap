// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VisualizationTypeListItem renders Column item 1`] = `
<div>
  <div
    class=""
  >
    <div>
      <svg
        style="width: 48px; height: 48px;"
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <polygon
            points="0 0 48 0 48 48 0 48"
          />
          <polygon
            fill="#147CD7"
            points="12 12 18 12 18 36 12 36"
          />
          <polygon
            fill="#147CD7"
            points="22 22 28 22 28 36 22 36"
          />
          <polygon
            fill="#147CD7"
            points="32 7 38 7 38 36 32 36"
          />
          <polygon
            fill="#4A5768"
            points="6 6 8 6 8 42 6 42"
          />
          <polygon
            fill="#4A5768"
            points="6 40 42 40 42 42 6 42"
          />
        </g>
      </svg>
    </div>
    <div>
      <p>
        Column Chart
      </p>
      <p>
        A chart that displays data in columns
      </p>
    </div>
  </div>
</div>
`;
